benchmark_runner/workloads/bootstorm_vm.py

```diff
@@ -173,99 +173,133 @@ def _verify_single_vm(self, vm_name, retries=5, delay=10):
-                    self.upload_run_artifacts_to_s3()
-                else:
-                    self._save_artifacts_local = True
+    def _verify_vms_in_parallel(self, vm_names):
+        """
+        This method verifies vms in parallel
+        :param vm_names:
+        :return:
+        """
+        failure_vms = []
+        manager = Manager()
+        return_dict = manager.dict()
+
+        # Split vm_names according to self._threads_limit
+        for i in range(0, len(vm_names), self._threads_limit):
+            bulk = vm_names[i:i + self._threads_limit]
+            processes = []
+
+            for vm_name in bulk:
+                p = Process(target=self._verify_single_vm_wrapper, args=(vm_name, return_dict))
+                p.start()
+                processes.append(p)
+
+            for p in processes:
+                p.join()
+
+        # After all processes are done, collect failures
+        for vm_name, status in return_dict.items():
+            if str(status).lower() != 'true':
+                failure_vms.append(vm_name)
+        return failure_vms
+
+    def _verify_virtctl_vms(self, delay=10):
+            """
+            This method verifies the virtctl SSH login for each VM, either during the upgrade or once for each VM.
+            It prepares the data for ElasticSearch, generates a must-gather in case of an error, and uploads it to Google Drive.
+            :param delay: delay between each iteration
+            """
+            try:
+                vm_names = self._oc._get_all_vm_names()
+                if not vm_names:
+                    raise MissingVMs("No VM names were retrieved from the cluster.")
 
-                # Error log with details of failed VM, for catching all vm errors
-                logger.error(
-                    f"Failed to verify virtctl SSH login for the following VMs: {', '.join(failure_vms)}")
+                upgrade_done = True
+                failure_vms = []  # List to store failed VM names
 
-        except Exception as err:
-            # Save run artifacts logs
-            self.save_error_logs()
-            raise err
+                if self._wait_for_upgrade_version:
+                    logger.info(f"wait for ocp upgrade version: {self._wait_for_upgrade_version}")
+                    upgrade_done = self._oc.get_cluster_status() == f'Cluster version is {self._wait_for_upgrade_version}'
+                    current_wait_time = 0
+
+                    while (self._timeout <= 0 or current_wait_time <= self._timeout) and not upgrade_done:
+                        failure_vms = self._verify_vms_in_parallel(vm_names)
+                        upgrade_done = self._oc.get_cluster_status() == f'Cluster version is {self._wait_for_upgrade_version}'
+
+                        # Sleep between each cycle
+                        time.sleep(delay)
+                        current_wait_time += delay  # Increment the wait time
```

Comment:
It should be something like this:
```
                    update_start_time = time.time()
                    while (self._timeout <= 0 or time.time() - update_start_time <= self._timeout) and not upgrade_done:
                        failure_vms = self._verify_vms_in_parallel(vm_names)
                        upgrade_done = self._oc.get_cluster_status() == f'Cluster version is {self._wait_for_upgrade_version}'

                        # Sleep between each cycle
                        time.sleep(delay)
```

Reply:
No need I have a timeout counter (up to 3600 sec), the start time is not relevant for timeout.

Reply:
You don't know how long the check takes.  What happens if it takes 20 seconds each time?  A minute?  Then the "3600 second" timeout is actually 10800 seconds.  Rather than assume that the current_wait_time actually increments by the delay, just compare the current time against the starting time.

Reply:
This is the way I have handled timeout in benchmark-runner.
`current_wait_time += delay `  it will failed when reaching to 3600, its default timeout.

Reply:
But you don't know what the *actual* delay is.  What if _verify_vms_in_parallel were to take 20 seconds, for example?

Reply:
Agree, Fixed

Reply:
Agree, Fixed